### PR TITLE
WICKET-6498 deferred javascript

### PR DESCRIPTION
**Pull request by svenmeier**
If addEventListener is used instead of jQuery  "domready" and "load" events, and OnLoadHeaderItem instead of JavaScriptHeaderItem, all JavaScript resources can be deferred.

Open questions:
- does this solution work reliably on all browsers?
- who decides which JavaScript resources to defer and which not?
- how to handle cases where user code is still using a JavaScriptHeaderItem?

### CODE REVIEW

File: wicket-core/src/main/java/org/apache/wicket/markup/head/OnLoadHeaderItem.java [hunk 1]
```diff
@@ -86,8 +86,8 @@ public void render(Response response)
 		CharSequence js = getJavaScript();
 		if (Strings.isEmpty(js) == false)
 		{
-			JavaScriptUtils.writeJavaScript(response, "Wicket.Event.add(window, \"load\", " +
-				"function(event) { " + js + ";});");
+			JavaScriptUtils.writeJavaScript(response,
+				"var f = function() {" + js + ";};\nif ('loaded' === document.readyState) f(); else window.addEventListener('load', f);");
```
Comment (klopfdreh) [inline wicket-core/src/main/java/org/apache/wicket/markup/head/OnLoadHeaderItem.java]: Is loaded the correct state for sure?

https://developer.mozilla.org/de/docs/Web/API/Document/readyState

here "complete" is mentioned. I also remember that there were some other pitfalls with the readyState for some browsers.

This german site mentions that "loaded" is only available in IE.

https://wiki.selfhtml.org/wiki/JavaScript/DOM/Document/readyState

File: wicket-core/src/main/java/org/apache/wicket/ajax/AbstractDefaultAjaxBehavior.java [hunk 2]
```diff
@@ -106,8 +107,7 @@ public void renderHead(final Component component, final IHeaderResponse response
 		RequestCycle requestCycle = component.getRequestCycle();
 		Url baseUrl = requestCycle.getUrlRenderer().getBaseUrl();
 		CharSequence ajaxBaseUrl = Strings.escapeMarkup(baseUrl.toString());
-		response.render(JavaScriptHeaderItem.forScript("Wicket.Ajax.baseUrl=\"" + ajaxBaseUrl
-			+ "\";", "wicket-ajax-base-url"));
```
Comment (klopfdreh) [inline wicket-core/src/main/java/org/apache/wicket/ajax/AbstractDefaultAjaxBehavior.java]: Mh I didn't look into the implementation but is wicket-ajax-base-url not used to be applied to the OnLoadHeaderItem, too. I could imagine that the information gets lost in this case.

File: wicket-core/src/main/java/org/apache/wicket/resource/CoreLibrariesContributor.java [hunk 3]
```diff
@@ -63,8 +64,7 @@ public static void contributeAjax(final Application application, final IHeaderRe
 		if (debugSettings.isAjaxDebugModeEnabled())
 		{
 			response.render(JavaScriptHeaderItem.forReference(jsLibrarySettings.getWicketAjaxDebugReference()));
-			response.render(JavaScriptHeaderItem.forScript("Wicket.Ajax.DebugWindow.enabled=true;",
-				"wicket-ajax-debug-enable"));
```
Comment (klopfdreh) [inline wicket-core/src/main/java/org/apache/wicket/resource/CoreLibrariesContributor.java]: Same as wicket-ajax-base-url

File: wicket-core/src/main/java/org/apache/wicket/ajax/AbstractDefaultAjaxBehavior.java (hunk 2 above)
Comment (klopfdreh) [reply wicket-core/src/main/java/org/apache/wicket/ajax/AbstractDefaultAjaxBehavior.java]: Ah this is only for the id attribute - so ignore this comment.

File: wicket-core/src/main/java/org/apache/wicket/resource/CoreLibrariesContributor.java (hunk 3 above)
Comment (klopfdreh) [reply wicket-core/src/main/java/org/apache/wicket/resource/CoreLibrariesContributor.java]: Also ignore this. :-)

File: wicket-core/src/main/java/org/apache/wicket/markup/head/OnLoadHeaderItem.java (hunk 1 above)
Comment (klopfdreh) [reply wicket-core/src/main/java/org/apache/wicket/markup/head/OnLoadHeaderItem.java]: Here I suggest to change it to „complete“

Comment (svenmeier) [reply wicket-core/src/main/java/org/apache/wicket/markup/head/OnLoadHeaderItem.java]: You're right, "complete" it should be.

File: wicket-core/src/main/java/org/apache/wicket/ajax/AbstractDefaultAjaxBehavior.java (hunk 2 above)
Comment (solomax) [reply wicket-core/src/main/java/org/apache/wicket/ajax/AbstractDefaultAjaxBehavior.java]: this missing `id` attribute breaks `org.apache.wicket.util.tester.BaseWicketTester.getWicketAjaxBaseUrlEncodedInLastResponse()` ....
Will set it back if there are no better solutions ...

Comment (svenmeier) [reply wicket-core/src/main/java/org/apache/wicket/ajax/AbstractDefaultAjaxBehavior.java]: It's not missing, it's just on a different location (see below), where getWicketAjaxBaseUrlEncodedInLastResponse() doesn't find it - we have to change that method accordingly.

Comment (solomax) [reply wicket-core/src/main/java/org/apache/wicket/ajax/AbstractDefaultAjaxBehavior.java]: previous code set the id attribute, so the check was trivial (node.attr === 'wicket-ajax-base-url'). Current implementation adds text content. which have to be parsed, maybe it worth to set attribute back to avoid parsing the text?

File: wicket-core/src/main/java/org/apache/wicket/markup/head/OnDomReadyHeaderItem.java [hunk 4]
```diff
@@ -86,8 +86,8 @@ public void render(Response response)
 		CharSequence js = getJavaScript();
 		if (Strings.isEmpty(js) == false)
 		{
-			JavaScriptUtils.writeJavaScript(response, "Wicket.Event.add(window, \"domready\", " +
-				"function(event) { " + js + ";});");
+			JavaScriptUtils.writeJavaScript(response,
+				"var f = function() {" + js + ";};\nif ('loading' !== document.readyState) f(); else document.addEventListener('DOMContentLoaded', f);");
```
Comment (solomax) [inline wicket-core/src/main/java/org/apache/wicket/markup/head/OnDomReadyHeaderItem.java]: I would like to change this code to be:
```
{ //scope
    const f = function() {" + js + ";}; ....
}
```

This way multiple `OnDomReadyHeader` items will work as expected in deferred mode
WDYT?

Comment (svenmeier) [reply wicket-core/src/main/java/org/apache/wicket/markup/head/OnDomReadyHeaderItem.java]: That does no harm +1

File: wicket-core/src/main/java/org/apache/wicket/ajax/AbstractDefaultAjaxBehavior.java (hunk 2 above)
Comment (svenmeier) [reply wicket-core/src/main/java/org/apache/wicket/ajax/AbstractDefaultAjaxBehavior.java]: Setting the baseUrl has to be deferred until onLoad, otherwise the deferred wicket-ajax-jquery.js hasn't been loaded yet and Wicket.Ajax is not defined.

File: wicket-core/src/main/java/org/apache/wicket/markup/head/ResourceAggregator.java [hunk 5]
```diff
@@ -337,24 +338,27 @@ private void renderCombinedEventScripts()
 		if (combinedScript.length() > 0)
 		{
 			combinedScript.append("\nWicket.Event.publish(Wicket.Event.Topic.AJAX_HANDLERS_BOUND);");
-			getRealResponse().render(
-				OnDomReadyHeaderItem.forScript(combinedScript.append('\n').toString()));
+			getRealResponse().render(OnDomReadyHeaderItem.forScript(combinedScript.append('\n')));
 		}
 
 		combinedScript.setLength(0);
+		String lastId = null;
```
Comment (solomax) [inline wicket-core/src/main/java/org/apache/wicket/markup/head/ResourceAggregator.java]: Would appreciate review on this commit, especially this part ....

Comment (svenmeier) [reply wicket-core/src/main/java/org/apache/wicket/markup/head/ResourceAggregator.java]: That doesn't seem to be reliable: what if multiple items are given an id?

IMHO we shouldn't changed code just to please test code. Couldn't BaseWicketTester just parse the url from the whole JavaScript via a regular expression?

> `.*Wicket\.Ajax\.baseUrl="(.*)";.*`

File: wicket-core/src/main/java/org/apache/wicket/markup/head/filter/FilteringHeaderResponse.java [hunk 6]
```diff
@@ -263,14 +262,12 @@ public void markRendered(Object object)
 			}
 		};
 
-		ResourceAggregator resourceAggregator = new ResourceAggregator(headerRenderer);
-
 		for (HeaderItem curItem : resp)
 		{
-			resourceAggregator.render(curItem);
+			headerRenderer.render(curItem);
```
Comment (solomax) [inline wicket-core/src/main/java/org/apache/wicket/markup/head/filter/FilteringHeaderResponse.java]: Unfortunately with this change `FilteringHeaderResponseTest` fails due to JS dependencies are not being rendered :(((
@svenmeier Could you please take a look at this?

Comment (svenmeier) [reply wicket-core/src/main/java/org/apache/wicket/markup/head/filter/FilteringHeaderResponse.java]: Thanks Maxim, I'll take a look.

**Comment by solomax** [comment]
Thanks a lot @svenmeier!
All tests are green now :)

Shall we merge?

**Comment by svenmeier** [comment]
Nope, not yet please. I have a new idea I want to try out.

**Comment by solomax** [comment]
Sure :) no rush, just would like to check if everyone happy with this PR :)

**Comment by svenmeier** [comment]
Many thanks Maxim for your work on this!

But with #258 we have a much simpler solution now, that should please those that are sceptical about the changes presented here.

If no one objects, I'll close this pull request.

**Comment by svenmeier** [comment]
@solomax I'm sorry for all the hard work you put into fixing the tests, but #258 is much better than this initial solution. Closing.

**Comment by solomax** [comment]
NP, I'll check the tests again  :)

**Comment by svenmeier** [comment]
There aren't many test to check - the new pull request does not change the current Ajax implementation at all :p. IMHO that's a huge advantage.
